### ISSUE TITLE
Verification Buffer is Extremely Small

### ISSUE THREAD

**Issue by lps-rocks** [transmission/transmission#1330]
https://github.com/transmission/transmission/blob/2ef07517b65e98c0c68bbff0ca2467434a1835d2/libtransmission/verify.c#L47

The verification buffer is hard-coded and extremely small - This should be a user configurable option to allow for faster hashing.

**Comment by ValZapod** [comment]
How faster it will be? It is already fast as hell here, 500 GB is hashed in 12 minutes.

**Comment by liut** [comment]
I have some large files that are over 400GB each, and the verification time is basically all over 20 hours, and I don't known why. Running on 32cores and 32G RAM, with 10T of disk space. Any idea?

**Comment by liut** [comment]
BTW. transmission(daemon) version is 3.0.

**Comment by GitHubGeek** [comment]
Similar issue here. Running on beefy Xeon server. Ubuntu 20.04 server. Transmission was compiled from source (090a4b5ff8f6b31a4eea4bdb75eb1350aa3efae0) 

Have this 900+GB torrent on verification for over 10 hours. Only 240 or so GB verified.

I feel this is a 3.0 specific issue, as past experience with 2.92 was very positive.

**Comment by liut** [comment]
BTW, my verifying logs:

...
[2021-03-12 18:06:29.051] A_FILE_WITH_122G Verified torrent piece #36 at 8 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:06:29.052] A_FILE_WITH_122G Verified torrent piece #37 at 8 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:06:29.052] A_FILE_WITH_122G Verified torrent piece #38 at 8 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:06:30.051] A_FILE_WITH_122G Verified torrent piece #39 at 9 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:06:30.051] A_FILE_WITH_122G Verified torrent piece #40 at 9 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:06:30.051] A_FILE_WITH_122G Verified torrent piece #41 at 9 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:06:30.051] A_FILE_WITH_122G Verified torrent piece #42 at 10 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:06:31.052] A_FILE_WITH_122G Verified torrent piece #43 at 10 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:06:31.052] A_FILE_WITH_122G Verified torrent piece #44 at 10 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:06:31.052] A_FILE_WITH_122G Verified torrent piece #45 at 10 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:06:31.052] A_FILE_WITH_122G Verified torrent piece #46 at 10 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:06:31.052] A_FILE_WITH_122G Verified torrent piece #47 at 10 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:06:32.052] A_FILE_WITH_122G Verified torrent piece #48 at 11 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:06:32.052] A_FILE_WITH_122G Verified torrent piece #49 at 11 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:06:32.052] A_FILE_WITH_122G Verified torrent piece #50 at 11 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:06:32.052] A_FILE_WITH_122G Verified torrent piece #51 at 11 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:06:32.052] A_FILE_WITH_122G Verified torrent piece #52 at 11 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:06:33.052] A_FILE_WITH_122G Verified torrent piece #53 at 12 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:06:47.052] A_FILE_WITH_122G Verified torrent piece #54 at 26 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:07:02.051] A_FILE_WITH_122G Verified torrent piece #55 at 41 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:07:16.052] A_FILE_WITH_122G Verified torrent piece #56 at 55 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:07:31.052] A_FILE_WITH_122G Verified torrent piece #57 at 71 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:07:45.052] A_FILE_WITH_122G Verified torrent piece #58 at 84 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:07:59.052] A_FILE_WITH_122G Verified torrent piece #59 at 98 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:08:14.052] A_FILE_WITH_122G Verified torrent piece #60 at 113 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:08:28.051] A_FILE_WITH_122G Verified torrent piece #61 at 127 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:08:43.052] A_FILE_WITH_122G Verified torrent piece #62 at 142 seconds (/app/libtransmission/verify.c:122)
[2021-03-12 18:08:57.051] A_FILE_WITH_122G Verified torrent piece #63 at 157 seconds (/app/libtransmission/verify.c:122)
...

**Comment by ValZapod** [comment]
> BTW, my verifying logs:

Your logs managed to pung all issues... Wow.

Haha

Anyway, what happens if you will increase the buffer? Is that really faster? Also, unteresting, what Tixati uses, since it is very fast.

**Comment by liut** [comment]
I tracked the most time consuming functions as follows:
```
verifyTorrent(tr_torrent* tor, bool* stopFlag)
 >> tr_torrentSetHasPiece(tr_torrent* tor, tr_piece_index_t pieceIndex, bool has)
  >> tr_cpPieceAdd(tr_completion* cp, tr_piece_index_t piece)
   >> tr_cpBlockAdd(tr_completion* cp, tr_piece_index_t piece, tr_block_index_t block)
    >> tr_bitfieldAdd(tr_bitfield* b, size_t nth)
```
The bitfiledAdd takes a few hundred milliseconds each time after a critical value is reached.

**Comment by liut** [comment]
build script at https://gist.github.com/liut/f137f15493876692b6e63a1c0e8befb3

**Comment by liut** [comment]
@ckerr I've added a patch.diff file to the above gist

**Comment by ckerr** [comment]
These assertions have been simplified in `main`. If the problem persists for anyone, please open a new issue with updated information. Thank you!

**Comment by lps-rocks** [comment]
@ckerr - Or how about we keep it open considering the buffer is still hard-coded and is abysmally small: 

https://github.com/transmission/transmission/blob/f415de0626cd88ad9b4cbedba10c53dfad0a24c8/libtransmission/verify.cc#L45

**Comment by ValZapod** [comment]
Does increasing that buffer that should be on the stack BTW really affects speed?

**Comment by lps-rocks** [comment]
> Does increasing that buffer that should be on the stack BTW really affects speed?

Yes, it does, drastically.

**Comment by ckerr** [comment]
> > Does increasing that buffer that should be on the stack BTW really affects speed?
> 
> Yes, it does, drastically.

I'm happy to spend more short-term memory use if the speed numbers warrant it. Do you have benchmarks?

**Comment by ckerr** [comment]
This issue is currently blocked on some numbers. @lps-rocks can you run some tests on the speed difference of different buffer sizes on your system?

**Comment by ile6695** [comment]
One interesting thing to check is if increasing buffer to some magic number like size of a file or L3 cache size would give the most benefit.
Increasing the buffer a lot could be implemented by keeping the currently working buffer size as is and catching an exception that gets thrown if using `std::reserve` for more memory fails. It has strong exception guarantee, so continuing normally would work in devices where there is not enough contiguous memory.

**Comment by ckerr** [comment]
If we use so much memory that we get a `std::length_error` thrown, we're doing something _very_ wrong :smile: 

I'm fine with the idea of increasing the buffer size, but I would like some objective numbers to know what sizes give more speedups and at what point we have diminishing returns. We _could_ just guess, but since some people in this thread have implied firsthand knowledge of different rates, it would be great to share that knowledge

**Comment by ckerr** [comment]
I was in the neighborhood of this issue while coding #2626 and tried tests where the buffer size was doubled, quadrupled, and then finally set to the piece size. I didn't see any change in speed -- in fact the larger buffer was slightly slower, though this is probably random chance.

It _is_ intuitive that a larger buffer size should speed things up, but it doesn't seem to bear out in practice, at least in my tests. Because of that, I'm going to close this issue.

If someone has a way to speed up the verify pass and the numbers to back it up, please open a new issue with details. Thank you!
